skaffold trace wrapping of critical functions & skaffold trace exporters via SKAFFOLD_TRACE env var

aaron-prindle opened the pull request:
### What is the problem being solved?
Part of #5756, adding opentelemetry trace information to skaffold commands. Added trace information to specific performance critical skaffold functions (identified in go/cloud-trace-skaffold).  Also added 4 trace exporters - gcp-skaffold, gcp-adc, stdout, and jaeger.  This PR uses env var based enabling/disabling for the trace for simplicity and to hide it from users directly.

### Why is this the best approach?
Using opentelemetry tracing is the obvious choice as we use open telemetry libs for metrics and it is becoming the metrics/tracing standard.  Using an env var in this PR and later integrating the flag setup was considered optimal as currently skaffold tracing will be used for benchmarking/bottleneck-identifying for select use cases while the user facing UX w/ jaeger, etc. is still being worked out.

### What other approaches did you consider?
There was the possibility of building tracing directly into skaffold events but I think with the current wrapper setup in `pkg/skaffold/instrumentation/trace.go` (w/ the minimal code required) and the fact that many trace locations will not be event locations  (eg: how long to hash a file, etc.) it makes sense to not integrate them.

### What side effects will this approach have?
There shouldn't be any side effects w/ this approach as the default "off" for tracing and the minimal user visibility for now should mean that it used only for select use cases experimentally.  I have done timing tests with the no-op/empty trace (`SKAFFOLD_TRACE` unset) and it does not change the performance of skaffold.

### What future work remains to be done?
Future work includes wiring up a --trace flag through dev, build, deploy, etc. and working on how skaffold might be able to do distributed tracing w/ other tools (minikube, buildpacks, etc.).  Additionally the ability to allow for more sporadic sampling (vs AlwaysSample) should be added.  Some future work mentioned in PR review included:
- OTEL_TRACES_EXPORTER=* support (vs SKAFFOLD_TRACE)
- Setup gcp-skaffold exporter and allow users to opt-in to emitting trace data at a sampling rate that does not affect performance

briandealwis commented on cmd/skaffold/app/cmd/runner.go:
```diff
@@ -64,13 +64,16 @@ func createNewRunner(out io.Writer, opts config.SkaffoldOptions) (runner.Runner,
 		return nil, nil, nil, fmt.Errorf("creating runner: %w", err)
 	}
-
+	instrumentation.InitMeterFromConfig(configs, opts.User)
+	_, _, err = instrumentation.InitTraceFromEnvVar()
+	if err != nil {
+		logrus.Debugf("error initializing tracing: %v", err)
+	}
```
Why shift this after `NewForConfig`?

aaron-prindle replied:
This was not intentional, moved both InitMeter* and InitTrace* to be before NewForConfig

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,37 @@
+In this example:
+
+* Use skaffold to deploy a jaeger local/remote jaeger instance
+* Enable skaffold tracing functionality to get trace information about skaffol dev, build, deploy, etc. timings
```
```suggestion
* Enable Skaffold tracing functionality to get trace information about skaffold `dev`, `build`, `deploy`, etc. timings
```

aaron-prindle replied:
Done

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,37 @@
+* Enable skaffold tracing functionality to get trace information about skaffol dev, build, deploy, etc. timings
+* Send skaffold trace information to jaeger and have that information be visible in the jaeger UI,
+
+In this example, we'll walk through enabling skaffold trace information that can be used to explore performance bottlenecks, etc. in skaffold and to get a more in depth view of user's local dev loop
```
```suggestion
In this example, we'll walk through enabling Skaffold trace information that can be used to explore performance bottlenecks and to get a more in depth view of user's local dev loop.
```

aaron-prindle replied:
Done

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,37 @@
+In this example, we'll walk through enabling skaffold trace information that can be used to explore performance bottlenecks, etc. in skaffold and to get a more in depth view of user's local dev loop
+
+**WARNING: If you're running this on a cloud cluster, this example will create a service and expose a webserver.
+It's highly suggested that you only run this example on a local, private cluster like minikube or Kubernetes in Docker for Desktop.**
```
```suggestion
It's highly recommended that you only run this example on a local, private cluster like minikube or Kubernetes in Docker for Desktop.**
```

aaron-prindle replied:
Done

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,37 @@
+
+In this example:
+
+* Use skaffold to deploy a jaeger local/remote jaeger instance
```
Jaeger and Skaffold should be capitalized since that's the conventional form both use.  Use `skaffold` to refer to the command.

```suggestion
* Use Skaffold to deploy a local/remote Jaeger instance
```

aaron-prindle replied:
Done

aaron-prindle replied:
Updated this in all relevant places for the example

briandealwis commented on examples/jaeger-skaffold-trace/jaeger-all-in-one-template.yaml:
```diff
@@ -0,0 +1,161 @@
+# Copyright 2019 The Skaffold Authors
```
Copyright should be 2021 on a new file.  But if you copied this from elsewhere then we need to maintain their copyright too.

aaron-prindle replied:
In my PR, the example I used for jaeger is from:
https://github.com/jaegertracing/jaeger-kubernetes/blob/master/all-in-one/jaeger-all-in-one-template.yml
(with small modifications - `s/extensions\/v1beta1/apps\/v1` & `s/port: 80/port: port: 16686`)

It is also Apache 2 licensed.

I have added an attribution in a comment at the top of the yaml.  Is there any specific way of attributing the authors there that makes sense?  I wasn't sure if the linters checked this exact blurb at the top.

aaron-prindle replied:
I've updated this to have:
- 2021
- an attribution to the original yaml file it is based off of

briandealwis commented on pkg/skaffold/build/cache/retrieve.go:
```diff
@@ -40,6 +41,9 @@ func (c *cache) Build(ctx context.Context, out io.Writer, tags tag.ImageTags, ar
 
 	start := time.Now()
 
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "CheckCache")
```
WDYT?
```suggestion
	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/build/cache", "CheckBuildCache")
```

aaron-prindle replied:
`CheckBuildCache` is more readable to me.  Done.

briandealwis commented on pkg/skaffold/graph/dependencies.go:
```diff
@@ -26,7 +26,10 @@ import (
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/instrumentation"
 	latestV1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v1"
+
+	// latest_v1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v1"
```
This line can be removed right?

aaron-prindle replied:
Yes, done.

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
````diff
@@ -0,0 +1,37 @@
+skaffold deploy --port-forward
+```
+
+Now, in a different terminal, go to another Skaffold example (eg: microservices), enable SKAFFOLD_TRACE w/ Jaeger and start dev session there:
````
```suggestion
Now, in a different terminal, go to another Skaffold example (eg: microservices), enable `SKAFFOLD_TRACE` with Jaeger and start dev session there:
```

aaron-prindle replied:
Done

briandealwis commented on examples/jaeger-skaffold-trace/jaeger-all-in-one-template.yaml:
```diff
@@ -0,0 +1,163 @@
+# Copyright 2021 The Skaffold Authors
```
Since you're copying this from elsewhere, we need to keep the original copyright intact.  Any changes we bring in are copyright by us.
```suggestion
# Copyright 2017-2019 The Jaeger Authors
# Copyright 2021 The Skaffold Authors
```

aaron-prindle replied:
Done

aaron-prindle replied:
> I think this is really cool.
> 
> Since this example isn't really standalone, what do you think about turning it into a tutorial instead? And rather than maintain the jaeger-all-in-one-yaml, why not use their Helm chart? https://github.com/jaegertracing/helm-charts
> 
> Remember to copy examples/ back into integration/examples/

The linked helm chart is for a production deployment of jaeger which might be overkill for what I was intending for usage.  We can discuss what makes more sense, I am keeping this as is for now.  There is a helm chart for the jaeger-all-in-one as well maintained by a solo contributor:
https://github.com/hansehe/jaeger-all-in-one

aaron-prindle replied:
Done

aaron-prindle replied:
Keeping the same setup for now then (example dir w/ README, k8s-yaml & skaffold yaml file)

briandealwis reviewed (commented):
I think this is really cool.

Since this example isn't really standalone, what do you think about turning it into a tutorial instead?  And rather than maintain the jaeger-all-in-one-yaml, why not use their Helm chart? https://github.com/jaegertracing/helm-charts

Remember to copy examples/ back into integration/examples/

briandealwis commented on pkg/skaffold/runner/v1/new.go:
```diff
@@ -89,6 +92,9 @@ func NewForConfig(runCtx *runcontext.RunContext) (*SkaffoldRunner, error) {
 	}
 
 	depLister := func(ctx context.Context, artifact *latestV1.Artifact) ([]string, error) {
+		ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "GetAllDependencies")
```
```suggestion
		ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "depLister")
```

It would be really useful to be able to include both the artifact name and type as attributes to we can see if there are particular artifacts that are bad.  Perhaps we could have an `instrumentation.Anonymize()` that can be toggled off (with an env var?).  We'll need to schedule a review (go/anon).

aaron-prindle replied:
Adding a stub `PII(a.ImageName)` fxn for now.  Made #5885 to track the future work here

briandealwis commented on pkg/skaffold/instrumentation/trace_test.go:
```diff
@@ -0,0 +1,159 @@
+				if err != nil {
+					t.Errorf("unexpected error occurred unmarshalling trace spans %v: %v", b.String(), err)
+				}
+				t.CheckTrue(len(spans) == len(test.spans))
```
Don't you want to compare the spans themselves?

aaron-prindle replied:
This part of the test verifies that when I define in the test I want X spans (X StartTrace calls) I indeed see X spans w/ the stdout exporter.  I could do a deep equal on the spans but I would need to mock more otel state and otel itself tests.  I am happy to make this change, I didn't know how useful it would be as it would mirror the tests done in otel itself:
https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/stdout/trace_test.go

briandealwis replied:
Maybe you could just compare what's output on `b`?  I'd like to verify a few situations: sequential spans are properly ordered, child spans are associated to a parent.  We could be messing up with the contexts, or how we call into the OpenTelemtry SDK.

aaron-prindle replied:
Added tests to verify all suggested cases:
sequential spans are properly ordered, concurrent spans work as expected, & child spans are associated to a parent

briandealwis commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -283,3 +299,160 @@ func errorMetrics(ctx context.Context, meter skaffoldMeter, m metric.Meter, labe
+		writer:    os.Stdout,
+		keyFS:     nil,
+		keyPath:   "/secret/keys.json",
+		jaegerURL: "http://localhost:14268/api/traces",
```
Can this be pushed into the `jaeger` section below?

aaron-prindle replied:
I have removed jaegerURL from the config here and hardcoded it directly into `initJaegerTraceExporter`

briandealwis commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -283,3 +299,160 @@ func errorMetrics(ctx context.Context, meter skaffoldMeter, m metric.Meter, labe
+		return initIOWriterTracer(teconf.writer)
+	}
+
+	switch os.Getenv("SKAFFOLD_TRACE") {
```
Can we use the `OTEL_TRACES_EXPORTER` environment variable? https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md#exporter-selection

aaron-prindle replied:
I am fine with changing our implementation to use OTEL_TRACES_EXPORTER vs SKAFFOLD_TRACE but I have some concerns:
1) It isn't clear to me if for example I use this env var w/ jaeger if that will interfere w/ the jaeger setup I have created (the url endpoint, etc.)
2) I think it is easier to rationalize when all skaffold env vars start w/ SKAFFOLD_* prefix, it might be confusing that skaffold uses this other env var which might be in the shell for other purposes/tools

Lmk if you prefer the OTEL_TRACES_EXPORTER env var and I can make the change.

briandealwis replied:
It looked like we'd get the same effect with `OTEL_TRACES_EXPORTER=jaeger` and `OTEL_EXPORTER_JAEGER_ENDPOINT=http://localhost:14268/`?

aaron-prindle replied:
These defaults are not working for me with the same setup I was using prior, I am not able to see any metrics in the jaeger UI using what I would think is the same setup via these env vars
```
OTEL_TRACES_EXPORTER=jaeger
OTEL_EXPORTER_JAEGER_ENDPOINT=http://localhost:14268/ # also tried http://localhost:14268/api/traces
```
I believe this will not work for the below reasons:

For the trace setup w/ jaeger we do create:
- a custom service "skaffold-trace"
- setup an sdk sampler w/ AlwaysSample

and in general we do the following:
- close all spans before os.Exit call via TracerShutdown callbacks

As such I will keep the SKAFFOLD_TRACE env var and recommend that be used.  Lmk and I can make the `OTEL_TRACES_EXPORTER` also make traceEnabled=true but currently it does not seem to work as I would expect

aaron-prindle replied:
I have made the OTEL_TRACES_EXPORTER env var make `traceEnabled=true` but I am having trouble verifying if they work properly

briandealwis replied:
My reasoning behind trying to leverage the OTEL*  variables is that we can then fit into existing systems.  We have some known partners (and a few unknown, I suspect) who may wish to integrate our traces with their own tracing.  They may be using something else other than Jaeger.

That said, the Go OpenTelemetry SDK for Go doesn't support OTEL_TRACES_EXPORTER (https://github.com/open-telemetry/opentelemetry-go/issues/1698).

So lets punt on this for now.

briandealwis commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -283,3 +299,160 @@ func errorMetrics(ctx context.Context, meter skaffoldMeter, m metric.Meter, labe
+		opt(&teconf)
+	}
+
+	if _, ok := os.LookupEnv("SKAFFOLD_EXPORT_TO_STDOUT"); ok {
+		logrus.Debugf("using stdout trace exporter")
+		return initIOWriterTracer(teconf.writer)
+	}
```
Do we need this if we have `SKAFFOLD_TRACE=stdout`?

aaron-prindle replied:
I have removed this now.  Metrics was using this variable `SKAFFOLD_EXPORT_TO_STDOUT` and it wasn't clear to me what the intent was so I duped the functionality here but agree the logic/usage is confusing atm.

briandealwis commented on pkg/skaffold/deploy/kubectl/kubectl.go:
```diff
@@ -92,15 +93,21 @@ func (k *Deployer) Deploy(ctx context.Context, out io.Writer, builds []graph.Art
 	// also, manually set the labels to ensure the runID is added
 	switch {
 	case len(k.hydratedManifests) > 0:
+		_, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/deploy/kubectl", "CreateManifestList")
```
So stylistic question: should we not have a trace for this entire `Deploy`, and then push these subtraces into the relevant functions?

aaron-prindle replied:
I have added this now.  My understanding was that the trace calls in applyResources:
```
func (r *SkaffoldRunner) applyResources(ctx context.Context, out io.Writer, artifacts, localImages []graph.Artifact) error {
	// Check that the cluster is reachable.
	// This gives a better error message when the cluster can't
	// be reached.
	if err := failIfClusterIsNotReachable(); err != nil {
		return fmt.Errorf("unable to connect to Kubernetes: %w", err)
	}

	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "LoadImagesIntoCluster")
	if len(localImages) > 0 && r.runCtx.Cluster.LoadImages {
		err := r.loadImagesIntoCluster(ctx, out, localImages)
		if err != nil {
			endTrace()
			return err
		}
	}
	endTrace()

	deployOut, postDeployFn, err := deployutil.WithLogFile(time.Now().Format(deployutil.TimeFormat)+".log", out, r.runCtx.Muted())
	if err != nil {
		return err
	}

	event.DeployInProgress()
	ctx, endTrace = instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DeployInProgress")
	defer endTrace()
	namespaces, err := r.deployer.Deploy(ctx, deployOut, artifacts)
```
were perhaps sufficient (following Event V1 API methods) but it makes sense to add tracing here as well.   My initial approach was to put trace statements where event statements were for the larger commands (dev, build, deploy, etc.) but it might make sense to wrap funcitons directly vs putting trace calls w/ Event API calls

aaron-prindle replied:
After talking offline I will keep the current traces in the place as Event Calls to help with possible correlation of the two later

briandealwis reviewed (commented):
The naming conventions could do with a revisit: the PR is often using the function name, but other times using something quite different.  I think we'd like to get some traceability back to the source to see where time is going?

WDYT of something like `"pkg", "Func[_Sub]"` where Func is the function name (e.g., `Dev`, or `doDev`, and `_Sub` is added if we're only tracing part of the function (e.g., `Dev_perArtifact`).

We also need to be very careful with PII in the attributes.  Let's chat about this offline.

aaron-prindle commented on pkg/skaffold/instrumentation/trace.go:
```diff
@@ -0,0 +1,74 @@
+	Value string
+}
+
+func StartTrace(ctx context.Context, pkg, name string, attributes ...TraceAttribute) (context.Context, func(options ...trace.SpanOption)) {
```
Added the go docs to the public commands, thanks.

I agree the current setup is confusing and verbose.  I have changed this to a variadic map that can be used similar to the third option:
```
func StartTrace(ctx context.Context, pkg, name string, attributes ...map[string]string)
```

I also like using the third option you listed:
```
func StartTrace(ctx context.Context, pkg, name string)
func StartTraceWithAttributes(ctx context.Context, pkg, name string, attributes map[string]string)
```
Lmk and I can change it from the variadic map version I currently changed it to (see above) to the 2 function version you posted.

aaron-prindle commented on pkg/skaffold/runner/v1/apply.go:
```diff
@@ -63,6 +64,8 @@ func (r *SkaffoldRunner) applyResources(ctx context.Context, out io.Writer, arti
 	}
 
 	event.DeployInProgress()
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DeployInProgress")
```
I have changed this to `Deploying`.  For trace events that also had actual Skaffold Events API methods I used the same name as the Events API method.

aaron-prindle replied:
Done

aaron-prindle commented on pkg/skaffold/runner/v1/apply.go:
```diff
@@ -63,6 +64,8 @@ func (r *SkaffoldRunner) applyResources(ctx context.Context, out io.Writer, arti
 	}
 
 	event.DeployInProgress()
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DeployInProgress")
```
Done

aaron-prindle commented on pkg/skaffold/runner/v1/deploy.go:
```diff
@@ -126,6 +127,9 @@ See https://skaffold.dev/docs/pipeline-stages/taggers/#how-tagging-works`)
 
 	event.DeployInProgress()
 	eventV2.TaskInProgress(constants.Deploy)
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DeployInProgress")
```
Done

aaron-prindle commented on pkg/skaffold/runner/v1/dev.go:
```diff
@@ -69,9 +69,11 @@ func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer, logger *logge
 	event.DevLoopInProgress(r.devIteration)
 	eventV2.TaskInProgress(constants.DevLoop)
 	defer func() { r.devIteration++ }()
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DevLoopInProgress")
```
Added devIteration attribute.  
Added doDev_needs[Sync|Build|Test|Deploy] parent traces here

aaron-prindle commented on pkg/skaffold/runner/v1/dev.go:
```diff
@@ -179,6 +182,8 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 	event.DevLoopInProgress(r.devIteration)
 	eventV2.TaskInProgress(constants.DevLoop)
 	defer func() { r.devIteration++ }()
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/runner", "DevLoopInProgress")
```
Done

aaron-prindle commented on pkg/skaffold/runner/v1/dev.go:
```diff
@@ -311,6 +316,7 @@ func (r *SkaffoldRunner) Dev(ctx context.Context, out io.Writer, artifacts []*la
 
 	event.DevLoopComplete(r.devIteration)
 	eventV2.TaskSucceeded(constants.DevLoop)
+	endTrace()
```
I have added endTrace() to all code paths now and added an attribute "error": "true" for the error return cases.  I've wired this through all other trace code as well.

aaron-prindle commented on pkg/skaffold/kubernetes/manifest/images.go:
```diff
@@ -58,7 +61,10 @@ func (is *imageSaver) Visit(o map[string]interface{}, k string, v interface{}) b
 // ReplaceImages replaces image names in a list of manifests.
-func (l *ManifestList) ReplaceImages(builds []graph.Artifact) (ManifestList, error) {
+func (l *ManifestList) ReplaceImages(ctx context.Context, builds []graph.Artifact) (ManifestList, error) {
+	_, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/kubernetes/manifest", "ReplaceImages")
```
Done

aaron-prindle commented:
> The naming conventions could do with a revisit: the PR is often using the function name, but other times using something quite different. I think we'd like to get some traceability back to the source to see where time is going?
> 
> WDYT of something like `"pkg", "Func[_Sub]"` where Func is the function name (e.g., `Dev`, or `doDev`, and `_Sub` is added if we're only tracing part of the function (e.g., `Dev_perArtifact`).
> 
> We also need to be very careful with PII in the attributes. Let's chat about this offline.

Regarding the naming convention, I have modified this to be of the recommended form.  Additionally I changed the pkg information to actually point to the file directly s/pkg\/skaffold\/runner/pkg\/skaffold\/runner\/v1\/deploy.go , etc.  Also in Cloud Trace currently the package information is not shown/used in the UI it seems so made it so the wrapper adds it as an attribute called `source_file`

briandealwis commented on pkg/skaffold/instrumentation/trace.go:
```diff
@@ -0,0 +1,85 @@
+// get parent/child traces a consistent context should be used.  The returned context will have the span metadata added so this context should
+// be used for subsequent calls in skaffold.  The returned function should be called to end the trace span, oftentimes this can be done with
+// the form:  _, endTrace = StartTrace...; defer endTrace()
+func StartTrace(ctx context.Context, pkg, name string, attributes ...map[string]string) (context.Context, func(options ...trace.SpanOption)) {
```
You could skip `pkg` and use Go's [`runtime.Caller()`](https://golang.org/pkg/runtime/#Caller) to get the file and location.  If there are any restrictions on the `name` you should mention it above.

aaron-prindle replied:
Removed `pkg` param and added `runtime.Caller()` value to be used instead for otel trace "pkg" param and the `source_file` attribute

briandealwis commented on pkg/skaffold/instrumentation/trace.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// StartTrace uses the otel trace provider to export timing spans (with optional attributes) to the chosen exporter
+//  via the value set in SKAFFOLD_TRACE.  Tracing is done via metadata stored in a context.Context, this means that to properly
```
```suggestion
// via the value set in SKAFFOLD_TRACE.  Tracing is done via metadata stored in a context.Context, this means that to properly
```

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/instrumentation/trace.go:
```diff
@@ -0,0 +1,85 @@
+
+// StartTrace uses the otel trace provider to export timing spans (with optional attributes) to the chosen exporter
+//  via the value set in SKAFFOLD_TRACE.  Tracing is done via metadata stored in a context.Context, this means that to properly
+// get parent/child traces a consistent context should be used.  The returned context will have the span metadata added so this context should
```
_consistent_ is a bit unclear?  How about just:
```suggestion
// get parent/child traces, callers should use the returned context 
```

aaron-prindle replied:
Done

aaron-prindle commented on pkg/skaffold/build/cache/hash.go:
```diff
@@ -82,6 +90,7 @@ func (h *artifactHasherImpl) hash(ctx context.Context, a *latestV1.Artifact) (st
 	if len(hashes) == 1 {
 		return hashes[0], nil
 	}
+
```
Not necessary, removed now

aaron-prindle commented on pkg/skaffold/build/scheduler.go:
```diff
@@ -90,11 +93,16 @@ func (s *scheduler) build(ctx context.Context, tags tag.ImageTags, i int) error
 	event.BuildInProgress(a.ImageName)
 	eventV2.BuildInProgress(i, a.ImageName)
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/build/scheduler.go", "build_BuildInProgress", map[string]string{
+		"ImageName": a.ImageName,
```
Done

aaron-prindle commented on pkg/skaffold/graph/dependencies.go:
```diff
@@ -57,13 +61,20 @@ type dependencyResolverImpl struct {
 
 func (r *dependencyResolverImpl) TransitiveArtifactDependencies(ctx context.Context, a *latestV1.Artifact) ([]string, error) {
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/graph/dependencies.go", "TransitiveArtifactDependencies", map[string]string{
+		"ArtifactName": a.ImageName,
```
Done

aaron-prindle commented on pkg/skaffold/graph/dependencies.go:
```diff
@@ -72,6 +83,11 @@ func (r *dependencyResolverImpl) TransitiveArtifactDependencies(ctx context.Cont
 
 func (r *dependencyResolverImpl) SingleArtifactDependencies(ctx context.Context, a *latestV1.Artifact) ([]string, error) {
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/graph/dependencies.go", "SingleArtifactDependencies", map[string]string{
+		"ArtifactName": a.ImageName,
```
Done

aaron-prindle commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -96,24 +104,33 @@ func exportMetrics(ctx context.Context, filename string, meter skaffoldMeter) er
+	return b, c.ProjectID, nil
+}
+
+func initCloudMonitoringExporterMetrics() (*basic.Controller, error) {
```
`initCloudMonitoringExporterMetrics` is the metrics exporter.  Initially I refactored this code a bit to share some of the statik/key logic for the similar trace exporter (gcp-skaffold).  As that exporter (gcp-skaffold) is removed now, I have undone the refactoring here as well.

aaron-prindle commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -283,3 +299,144 @@ func errorMetrics(ctx context.Context, meter skaffoldMeter, m metric.Meter, labe
+	teconf := TraceExporterConfig{
+		writer:  os.Stdout,
+		keyFS:   nil,
+		keyPath: "/secret/keys.json",
```
Done

aaron-prindle commented on pkg/skaffold/instrumentation/export.go:
```diff
@@ -283,3 +299,144 @@ func errorMetrics(ctx context.Context, meter skaffoldMeter, m metric.Meter, labe
+			}
+			teconf.keyFS = statikFS
+		}
+		return initCloudTraceExporterSkaffold(teconf.keyFS, teconf.keyPath)
```
Done

aaron-prindle commented:
Adding a comment here as a TODO regarding using build tags and adding binary size information in this PR

aaron-prindle commented on pkg/skaffold/build/cache/hash.go:
```diff
@@ -66,14 +69,19 @@ func newArtifactHasher(artifacts graph.ArtifactGraph, lister DependencyLister, m
+
 	hash, err := h.safeHash(ctx, a)
 	if err != nil {
+		endTrace(trace.WithAttributes(attribute.KeyValue{Key: "error", Value: attribute.StringValue("true")}))
```
Done.  Now this has the form:
endTrace(WithError(err))

aaron-prindle commented on pkg/skaffold/graph/dependencies.go:
```diff
@@ -72,6 +77,10 @@ func (r *dependencyResolverImpl) TransitiveArtifactDependencies(ctx context.Cont
 
 func (r *dependencyResolverImpl) SingleArtifactDependencies(ctx context.Context, a *latestV1.Artifact) ([]string, error) {
+	ctx, endTrace := instrumentation.StartTrace(ctx, "pkg/skaffold/graph", "SingleArtifactDependencies",
+		instrumentation.TraceAttribute{Key: "ArtifactName", Value: a.ImageName})
```
Done, added PII(a.ImageName) stub to mark these for now

briandealwis commented:
It would be helpful to add some tracing to the rendering/deploy steps — @dhodun is reporting that this can take 30-45 seconds on his setup with 15 containers.

aaron-prindle commented:
It seems that while skaffold v1.24.0 is ~53 mb, skaffold @ HEAD is ~60mb.  My PR actually changes the size of skaffold negligibly:
60->61mb

aaron-prindle commented:
Does it still make sense to gaurd this with build tags given this?  Currently I have split this out in a separate branch w/

// +build profile

// +build !profile

but it seems `golangci-lint` is not happy with the build tags:
```
aprindle@aprindle ~/skaffold  [skaff-trace]$ make linters
Running linters...
RUN hack/boilerplate.sh
PASSED hack/boilerplate.sh in 0s
RUN hack/gofmt.sh
PASSED hack/gofmt.sh in 1s
RUN hack/golangci-lint.sh
WARN [runner] Can't run linter goanalysis_metalinter: bodyclose: failed prerequisites: [buildssa@github.com/GoogleContainerTools/skaffold/pkg/skaffold/instrumentation [github.com/GoogleContainerTools/skaffold/pkg/skaffold/instrumentation.test]: analysis skipped: errors in package: [/usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:29:6: TestInitCloudTrace redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:29:6: 	other declaration of TestInitCloudTrace /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:73:6: SpanArray redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:73:6: 	other declaration of SpanArray /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:92:6: Spancontext redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:92:6: 	other declaration of Spancontext /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:99:6: Parent redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:99:6: 	other declaration of Parent /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:106:6: Value redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:106:6: 	other declaration of Value /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:110:6: Resource redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:110:6: 	other declaration of Resource /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:114:6: Instrumentationlibrary redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:114:6: 	other declaration of Instrumentationlibrary /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:52:39: undeclared name: WithWriter /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:52:39: undeclared name: WithWriter]] 
WARN [runner] Can't run linter unused: buildir: analysis skipped: errors in package: [/usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:29:6: TestInitCloudTrace redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:29:6: other declaration of TestInitCloudTrace /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:73:6: SpanArray redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:73:6: 	other declaration of SpanArray /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:92:6: Spancontext redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:92:6: 	other declaration of Spancontext /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:99:6: Parent redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:99:6: 	other declaration of Parent /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:106:6: Value redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:106:6: 	other declaration of Value /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:110:6: Resource redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:110:6: 	other declaration of Resource /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:114:6: Instrumentationlibrary redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:114:6: 	other declaration of Instrumentationlibrary /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:52:39: undeclared name: WithWriter /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:52:39: undeclared name: WithWriter] 
ERRO Running error: buildir: analysis skipped: errors in package: [/usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:29:6: TestInitCloudTrace redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:29:6: 	other declaration of TestInitCloudTrace /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:73:6: SpanArray redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:73:6: 	other declaration of SpanArray /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:92:6: Spancontext redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:92:6: 	other declaration of Spancontext /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:99:6: Parent redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:99:6: 	other declaration of Parent /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:106:6: Value redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:106:6: 	other declaration of Value /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:110:6: Resource redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:110:6: 	other declaration of Resource /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:114:6: Instrumentationlibrary redeclared in this block /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:114:6: 	other declaration of Instrumentationlibrary /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_profile_test.go:52:39: undeclared name: WithWriter /usr/local/google/home/aprindle/skaffold/pkg/skaffold/instrumentation/trace_test.go:52:39: undeclared name: WithWriter] 
FAILED hack/golangci-lint.sh
```

even though it compiles-correctly/is-correct.  I haven't found the correct ignore config for `golangci.yaml`.  Given the negligible binary size difference do we still want to pursue this?

briandealwis commented on pkg/skaffold/build/cache/hash.go:
```diff
@@ -66,14 +67,19 @@ func newArtifactHasher(artifacts graph.ArtifactGraph, lister DependencyLister, m
 }
 
 func (h *artifactHasherImpl) hash(ctx context.Context, a *latestV1.Artifact) (string, error) {
+	ctx, endTrace := instrumentation.StartTrace(ctx, "hash_GenerateHashOneArtifact")
```
This seems like a good place to include the image name
```suggestion
	ctx, endTrace := instrumentation.StartTrace(ctx, "hash_GenerateHashOneArtifact", map[string]string{
		"ImageName": instrumentation.PII(a.ImageName),
	}))
```

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/build/cache/lookup.go:
```diff
@@ -50,6 +54,9 @@ func (c *cache) lookupArtifacts(ctx context.Context, tags tag.ImageTags, artifac
 }
 
 func (c *cache) lookup(ctx context.Context, a *latestV1.Artifact, tag string, h artifactHasher) cacheDetails {
+	ctx, endTrace := instrumentation.StartTrace(ctx, "lookup_CacheLookupOneArtifact")
```
WDYT?
```suggestion
	ctx, endTrace := instrumentation.StartTrace(ctx, "lookup_CacheLookupOneArtifact", map[string]string{
		"ImageName": instrumentation.PII(a.ImageName),
	})))
```

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/build/scheduler.go:
```diff
@@ -90,11 +91,16 @@ func (s *scheduler) build(ctx context.Context, tags tag.ImageTags, i int) error
 
 	event.BuildInProgress(a.ImageName)
 	eventV2.BuildInProgress(i, a.ImageName)
+	ctx, endTrace := instrumentation.StartTrace(ctx, "build_BuildInProgress", map[string]string{
+		"ImageName": instrumentation.PII(a.ImageName),
+	})
```
Can we incorporate the builder type?  And the artifact number `i`?

aaron-prindle replied:
Is the builder type accessible in `scheduler.go` here?  Current approach would be to do something like adding attributes to the current trace (via otel [SpanFromContext](https://github.com/opentracing/opentracing-go/blob/3088eee7e4d26010ac9a301e977fd2721dbebcbe/gocontext.go#L20-L32))

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/deploy/kubectl/kubectl.go:
```diff
@@ -88,43 +89,62 @@ func (k *Deployer) Deploy(ctx context.Context, out io.Writer, builds []graph.Art
 		err       error
 	)
+
+	ctx, endTrace := instrumentation.StartTrace(ctx, "Deploy")
```
This should indicate the deployer type. WDYT:
```suggestion
	ctx, endTrace := instrumentation.StartTrace(ctx, "deploy_kubectl")
```

Or should it be an attribute?

aaron-prindle replied:
Done.  Added as an attribute

briandealwis commented on pkg/skaffold/deploy/kpt/kpt.go:
```diff
@@ -375,7 +375,7 @@ func (k *Deployer) renderManifests(ctx context.Context, _ io.Writer, builds []gr
 	if err != nil {
```
Can you add some tracing for `Deploy`, `Render`, and `Apply` as you've done for the kubectl deployer?

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/deploy/kustomize/kustomize.go:
```diff
@@ -189,7 +189,7 @@ func (k *Deployer) renderManifests(ctx context.Context, out io.Writer, builds []
 		return nil, nil
 	}
```
It would be useful to have similar tracing in the kustomize deployer.

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/runner/v1/new.go:
```diff
@@ -79,31 +84,38 @@ func NewForConfig(runCtx *runcontext.RunContext) (*SkaffoldRunner, error) {
 	}
 
 	depLister := func(ctx context.Context, artifact *latestV1.Artifact) ([]string, error) {
+		ctx, endTrace := instrumentation.StartTrace(ctx, "GetAllDependencies")
```
There isn't a method GetAllDependencies — how about `NewForConfig_depLister`?

aaron-prindle replied:
Done

briandealwis reviewed (commented):
This is looking good — the TraceEndError is much more readable.

aaron-prindle commented on cmd/skaffold/app/cmd/runner.go:
```diff
@@ -65,12 +65,15 @@ func createNewRunner(out io.Writer, opts config.SkaffoldOptions) (runner.Runner,
 	}
 
 	instrumentation.InitMeterFromConfig(configs, opts.User)
+	_, _, err = instrumentation.InitTraceFromEnvVar()
+	if err != nil {
+		logrus.Debugf("error initializing tracing: %v", err)
+	}
```
Done.  Moved the init-failure logging into the function, similar to InitMeterFromConfig() and merged instrumentation.InitMeter/InitTrace to new instrumentation.Init

aaron-prindle commented on cmd/skaffold/skaffold.go:
```diff
@@ -49,6 +49,10 @@ func main() {
 	if err := instrumentation.ExportMetrics(code); err != nil {
 		logrus.Debugf("error exporting metrics %v", err)
 	}
+	if err := instrumentation.TracerShutdown(context.Background()); err != nil {
+		logrus.Debugf("error shutting down tracer %v", err)
+	}
+
```
Merged both shutdown/flush methods for metrics&trace into `ShutdownAndFlush`

aaron-prindle commented on examples/jaeger-skaffold-trace/jaeger-all-in-one-template.yaml:
```diff
@@ -0,0 +1,164 @@
+# limitations under the License.
+
+
+# adapted from https://github.com/jaegertracing/jaeger-kubernetes/blob/master/all-in-one/jaeger-all-in-one-template.yml
```
Added a blurb with the following:
```
# adapted from https://github.com/jaegertracing/jaeger-kubernetes/blob/master/all-in-one/jaeger-all-in-one-template.yml
# changes made include:
#   - updating Deployment to correct apiVersion for newer k8s API - extensions/v1beta1 ->  apps/v1
#   - fixing deployment to have required `selector` field in newer k8s Deployment API
#   - changed Service - query-http to not use port 80 -> port 16686 which is the export local jaeger port (vs web hosted)   
```

briandealwis commented on pkg/skaffold/deploy/kubectl/kubectl.go:
```diff
@@ -110,21 +122,27 @@ func (k *Deployer) Deploy(ctx context.Context, out io.Writer, builds []graph.Art
 	}
+	endTrace()
 
+	ctx, endTrace = instrumentation.StartTrace(ctx, "Deploy_WaitForDeletions")
```
Do we need to worry that `ctx` (the argument) is actually the ctx from the previous `StartTrace()`, and not the ctx passed into this function?  Does the `endTrace()` re-link the context?

aaron-prindle replied:
I have changed the `Deploy_` and `doDev_` calls which had multiple `StartTrace` calls in a single function and passed `ctx` down the chain to properly create a new ctx (`childCtx`) and pass that to the relevant children for each command s.t. each direct trace in the function uses the initial root ctx

briandealwis commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,37 @@
+### Example: Skaffold Command Tracing with Jaeger
+
```
Can you put a note here that this is experimental and may change without notice.  And be sure to place it in integration/examples too

aaron-prindle replied:
Done

aaron-prindle commented on examples/jaeger-skaffold-trace/README.md:
```diff
@@ -0,0 +1,40 @@
+
+In this example, we'll walk through enabling Skaffold trace information that can be used to explore performance bottlenecks and to get a more in depth view of user's local dev loop.
+
+_**WARNING: If you're running this on a cloud cluster, this example will create a service and expose a webserver.
+It's highly recommended that you only run this example on a local, private cluster like minikube or Kubernetes in Docker for Desktop.**_
```
Done

aaron-prindle commented on pkg/skaffold/build/docker/docker.go:
```diff
@@ -24,13 +24,18 @@ import (
 
 func (b *Builder) Build(ctx context.Context, out io.Writer, a *latestV1.Artifact, tag string) (string, error) {
+	instrumentation.AddAttributesToCurrentSpanFromContext(ctx, map[string]string{
+		"BuildType": "docker",
```
Done

aaron-prindle commented on pkg/skaffold/deploy/kpt/kpt.go:
```diff
@@ -132,41 +133,67 @@ func versionCheck(dir string, stdout io.Writer) error {
+	endTrace()
+
+	_, endTrace = instrumentation.StartTrace(ctx, "Deploy_execKptCommand")
+	cmd := exec.CommandContext(childCtx, "kpt", kptCommandArgs(applyDir, []string{"live", "apply"}, k.getKptLiveApplyArgs(), nil)...)
```
Thanks for catching this, this was a mistake.  It now uses the context ignored on 187 as was originally intended.

briandealwis commented on pkg/skaffold/build/custom/build.go:
```diff
@@ -22,11 +22,17 @@ import (
+	instrumentation.AddAttributesToCurrentSpanFromContext(ctx, map[string]string{
+		"BuildType":   "custom",
+		"Context":     instrumentation.PII(artifact.Workspace),
+		"Destination": tag,
```
```suggestion
		"Destination": instrumentation.PII(tag),
```

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/build/docker/docker.go:
```diff
@@ -24,13 +24,20 @@ import (
+	instrumentation.AddAttributesToCurrentSpanFromContext(ctx, map[string]string{
+		"BuildType":   "docker",
+		"Context":     instrumentation.PII(a.Workspace),
+		"Destination": tag,
```
```suggestion
		"Destination": instrumentation.PII(tag),
```

aaron-prindle replied:
Done

briandealwis commented on pkg/skaffold/build/jib/build.go:
```diff
@@ -20,11 +20,18 @@ import (
+	instrumentation.AddAttributesToCurrentSpanFromContext(ctx, map[string]string{
+		"BuildType":   "jib",
+		"Context":     instrumentation.PII(artifact.Workspace),
+		"Destination": tag,
```
```suggestion
		"Destination": instrumentation.PII(tag),
```

aaron-prindle replied:
Done

briandealwis reviewed (approved):
I think the destination tags should be marked as PII.  With that, LGTM.

aaron-prindle commented on pkg/skaffold/build/gcb/cloud_build.go:
```diff
@@ -62,6 +67,9 @@ func (b *Builder) Concurrency() int {
 
 func (b *Builder) buildArtifactWithCloudBuild(ctx context.Context, out io.Writer, artifact *latestV1.Artifact, tag string) (string, error) {
+	instrumentation.AddAttributesToCurrentSpanFromContext(ctx, map[string]string{
+		"Destination": tag,
```
Done
